cloudera/cm_csds
# Add Zeppelin CSD

**What is this PR for?**
This PR is for creating a Zeppelin Custom Service Descriptor (CSD) for Cloudera Manager (CM) to support the deployment and management of Zeppelin on Cloudera clusters. This change is necessary because, starting from version CDH 7.3.1, Zeppelin will be disabled by default. The PR involves downloading the necessary code, integrating the Zeppelin CSD, building the JAR file, and deploying it on a Cloudera cluster.

**What is the Jira issue?**
[CDPD-71714](https://jira.cloudera.com/browse/CDPD-71714) (Re-enabling Zeppelin in 7.3.1+ via external CSD)

**Questions:**
_Are there breaking changes for older versions?_  No
_Does this need documentation?_  Yes

**How pre-commit build is tested?**
There is no pre-commit build for this repository. So, it is tested locally with below configuration:
_Pre requisite:_
MVN: 3.8.8
Java: 1.8

_Maven build:_
```
$ mvn clean install
...
...
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO] 
[INFO] The Enterprise CSDs 5.14.0 ......................... SUCCESS [  0.309 s]
[INFO] The Accumulo CSD 5.14.0 ............................ SUCCESS [  6.205 s]
[INFO] The Accumulo 1.6 CSD 5.14.0 ........................ SUCCESS [  4.298 s]
[INFO] The Echo Service CSD 5.14.0 ........................ SUCCESS [  3.124 s]
[INFO] Kafka CSD 5.14.0 ................................... SUCCESS [ 16.441 s]
[INFO] The Cloudera Navigator Key Trustee CSD 5.14.0 ...... SUCCESS [  4.133 s]
[INFO] The Hadoop Key Management Server CSD 5.14.0 ........ SUCCESS [  3.645 s]
[INFO] The Spark CSD 5.14.0 ............................... SUCCESS [  4.438 s]
[INFO] The Spark on YARN CSD 5.14.0 ....................... SUCCESS [  4.704 s]
[INFO] The Spark on YARN 53 CSD 5.14.0 .................... SUCCESS [  4.332 s]
[INFO] The Spark on YARN 54 CSD 5.14.0 .................... SUCCESS [  4.821 s]
[INFO] The Spark on YARN 5.10 CSD 5.14.0 .................. SUCCESS [  4.490 s]
[INFO] The Sqoop Client CSD 5.14.0 ........................ SUCCESS [  2.467 s]
[INFO] Zeppelin CSD for CDH 7.3.1+ 7.13.1 ................. SUCCESS [  3.200 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  01:06 min
[INFO] Finished at: 2024-08-23T11:44:59+05:30
[INFO] ------------------------------------------------------------------------
```


**How should this be tested?**
1. Follow the below commands to generate a CSD jar and copy it to CM cluster:
```
$cd cm_csds
$mvn clean install
$ls -lrt ZEPPELIN/target/ZEPPELIN-7.13.1.jar
$scp ZEPPELIN/target/ZEPPELIN-7.13.1.jar <user>@<CM host>:/opt/cloudera/csd/
$ssh <user>@<host>
$cd /opt/cloudera/csd/
$sudo chmod 664 ZEPPELIN-7.13.1.jar
$sudo chown cloudera-scm:cloudera-scm ZEPPELIN-7.13.1.jar
$sudo service cloudera-scm-server restart
```

2. Open the Cloudera Manager Admin Console.
![Screenshot 2024-08-09 at 5 38 42 PM](https://github.com/user-attachments/assets/f657b54f-71a0-4255-9dce-6d9cb5b0620b)

3. Navigate to Cloudera Management Service and restart any stale services.
<img width="1117" alt="Screenshot 2024-08-23 at 12 20 10 PM" src="https://github.com/user-attachments/assets/f4aa598f-45c6-4012-896d-938a23e10a51">

4. Go to CM Home -> Status -> Click on the settings or the 3 dots on the right side of the relevant cluster. -> Select Add Service.
![Screenshot 2024-08-09 at 5 38 49 PM](https://github.com/user-attachments/assets/bccb4dd3-451b-4d0c-b61e-91b5f9d5092a)

5. Select Zeppelin.
![Screenshot 2024-08-09 at 5 39 14 PM](https://github.com/user-attachments/assets/6594578d-bd50-4af2-8903-ef76d22b991d)

6. Follow the instructions to add the Zeppelin service to your cluster.
![Screenshot 2024-08-09 at 5 45 12 PM](https://github.com/user-attachments/assets/754cdc3e-6ef6-42a8-9fa7-73c0fdb5a8c3)

7. Start Zeppelin
![Screenshot 2024-08-09 at 5 45 38 PM](https://github.com/user-attachments/assets/83b1f256-5c95-4f76-b914-f9b1e1792e97)

8. Open Zeppelin in the browser:
![Screenshot 2024-08-09 at 5 48 11 PM](https://github.com/user-attachments/assets/2ff6c41d-c1f2-4600-b5b2-d48fb4167c75)

9. Test with any code and see the output to confirm it is working as expected.
![Screenshot 2024-08-09 at 5 52 33 PM](https://github.com/user-attachments/assets/9ee74688-89ce-4908-a528-27b90ddc5c28)
